From 11ebc02d3914b5b3e05c964ac1d21f7e5b8cd6be Mon Sep 17 00:00:00 2001
From: yuayu <naoki66@163.com>
Date: Fri, 16 Oct 2026 15:47:26 +0800
Subject: [PATCH] miniupnpd: refresh patches for 2.3.11

- drop 020-descr-filter-fix.patch: the iseol/get_sep permission parser
  rewrite it carried is already present in the 2.3.11 tarball.
- drop 030-allow-private-fix.patch: the ALLOWPRIVATEIPV4MASK logic fix it
  carried is already present in the 2.3.11 tarball.
- refresh 040-improve-logging.patch against the 2.3.11 sources: the usage
  text block in miniupnpd.c moved, so several hunks no longer applied.
---
 .../patches/020-descr-filter-fix.patch        | 511 ------------------
 .../patches/030-allow-private-fix.patch       |  18 -
 .../patches/040-improve-logging.patch         | 184 +++----
 3 files changed, 92 insertions(+), 621 deletions(-)
 delete mode 100644 net/miniupnpd/patches/020-descr-filter-fix.patch
 delete mode 100644 net/miniupnpd/patches/030-allow-private-fix.patch

diff --git a/net/miniupnpd/patches/020-descr-filter-fix.patch b/net/miniupnpd/patches/020-descr-filter-fix.patch
deleted file mode 100644
index 9d657d1534..0000000000
--- a/net/miniupnpd/patches/020-descr-filter-fix.patch
+++ /dev/null
@@ -1,511 +0,0 @@
-From 54ecff0ae4452598773a020e8798ff61ccf3c966 Mon Sep 17 00:00:00 2001
-From: yangfl <yangfl@users.noreply.github.com>
-Date: Thu, 31 Jul 2025 10:35:50 +0800
-Subject: [PATCH] miniupnpd: Rewrite permission line parser
-
-Permission line parser rewrites input buffer in-place for parsing,
-which causes several problems:
-
-* It unnecessarily invalidates input buffer for caller. Actually, you
-  might see the following error message when parsing fails:
-    miniupnpd[1234]: parsing error file /etc/miniupnpd.conf line 16 : allow 1024
-  since the hyphen after 1024 is erased by a '\0'.
-
-* It fails to validate token separators. For example, the following
-  line will be accepted:
-    allow 1024-65535X0.0.0.0/0 1024-65535 all
-                    ^ could be any character
-  and even a potential buffer over-read if the character is '\0', since
-  the parser blindlessly skips it.
-
-* The fifth token is never parsed since it gets a previously written
-  '\0'.
-
-Instead of fixing them case-by-case, we rewrite it with DFS in mind,
-which is much simpler and less error-prone.
-
---- a/upnppermissions.c
-+++ b/upnppermissions.c
-@@ -29,6 +29,12 @@ isodigit(char c)
- 	return '0' <= c && c >= '7';
- }
- 
-+static int
-+iseol(char c)
-+{
-+	return c == '\0' || c == '\n' || c == '\r';
-+}
-+
- static char
- hex2chr(char c)
- {
-@@ -91,6 +97,84 @@ unescape_char(const char * s, int * seql
- 	return c;
- }
- 
-+/* greedy parser: try to match the longest sequence and do not
-+ * check for terminators */
-+
-+static char *
-+get_sep(const char * s)
-+{
-+	if(!isspace(*s))
-+		return NULL;
-+	do
-+		s++;
-+	while(isspace(*s));
-+	return (char *) s;
-+}
-+
-+static char *
-+get_ushort(const char * s, u_short * val)
-+{
-+	char * end;
-+	unsigned long val_ul;
-+
-+	if(!isdigit(*s))
-+		return NULL;
-+	val_ul = strtoul(s, &end, 10);
-+	if(val_ul > 65535)
-+		return NULL;
-+	*val = (u_short)val_ul;
-+
-+	return end;
-+}
-+
-+static char *
-+get_range(const char * s, u_short * begin, u_short * end)
-+{
-+	s = get_ushort(s, begin);
-+	if(!s)
-+		return NULL;
-+
-+	if(*s!='-')
-+		*end = *begin;
-+	else
-+	{
-+		s++;
-+		s = get_ushort(s, end);
-+		if(!s)
-+			return NULL;
-+		if(*begin > *end)
-+			return NULL;
-+	}
-+	return (char *) s;
-+}
-+
-+static char *
-+get_addr(const char * s, struct in_addr * addr, unsigned int * dot_cnt)
-+{
-+	size_t i;
-+	char buf[64];
-+
-+	if(!isdigit(*s))
-+		return NULL;
-+
-+	*dot_cnt = 0;
-+	for(i = 0; isdigit(s[i]) || s[i] == '.';)
-+	{
-+		if(s[i] == '.')
-+			(*dot_cnt)++;
-+		buf[i] = s[i];
-+		i++;
-+		if (i > sizeof(buf) - 1)
-+			return NULL;
-+	}
-+
-+	buf[i] = '\0';
-+	if(!inet_aton(buf, addr))
-+		return NULL;
-+
-+	return (char *) s + i;
-+}
-+
- /* get_next_token(s, &token, raw)
-  * put the unquoted/unescaped token in token and returns
-  * a pointer to the begining of the next token
-@@ -99,18 +183,8 @@ static char *
- get_next_token(const char * s, char ** token, int raw)
- {
- 	char deli;
--	const char * end;
-+	size_t len;
- 
--	/* skip any whitespace */
--	for(; isspace(*s); s++)
--		if(*s == '\0' || *s == '\n')
--		{
--			if(token)
--				*token = NULL;
--			return (char *) s;
--		}
--
--	/* find the start */
- 	if(*s == '"' || *s == '\'')
- 	{
- 		deli = *s;
-@@ -119,85 +193,90 @@ get_next_token(const char * s, char ** t
- 	else
- 		deli = 0;
- 	/* find the end */
--	end = s;
--	for(; *end != '\0' && *end != '\n' && (deli ? *end != deli : !isspace(*end));
--	    end++)
--		if(*end == '\\')
-+	for(len = 0; !iseol(s[len]) && (deli ? s[len] != deli : !isspace(s[len]));
-+	    len++)
-+		if(s[len] == '\\')
- 		{
--			end++;
--			if(*end == '\0')
-+			len++;
-+			if(iseol(s[len]))
- 				break;
- 		}
- 
- 	/* save the token */
- 	if(token)
- 	{
--		unsigned int token_len;
--		unsigned int i;
--
--		token_len = end - s;
--		*token = strndup(s, token_len);
--		if(!*token)
--			return NULL;
--
--		for(i = 0; (*token)[i] != '\0'; i++)
--		{
--			int sequence_len;
--
--			if((*token)[i] != '\\')
--				continue;
--
--			if(raw && deli && (*token)[i + 1] != deli)
--				continue;
--			(*token)[i] = unescape_char(*token + i, &sequence_len);
--			memmove(*token + i + 1, *token + i + sequence_len,
--			        token_len - i - sequence_len);
--		}
--		if (i == 0)
--		{
--			/* behavior of realloc(p, 0) is implementation-defined, so better set it to NULL.
--			 * https://github.com/miniupnp/miniupnp/issues/652#issuecomment-1518922139 */
--			free(*token);
-+		if(len == 0)
- 			*token = NULL;
--		}
- 		else
- 		{
--			char * tmp = realloc(*token, i);
--			if (tmp != NULL)
--				*token = tmp;
-+			unsigned int i;
-+			unsigned int j;
-+
-+			char * tmp;
-+			char * t;
-+
-+			t = malloc(len + 1);
-+			if(!t)
-+				return NULL;
-+
-+			if (raw)
-+			{
-+				memcpy(t, s, len);
-+				j = len;
-+			}
- 			else
--				syslog(LOG_ERR, "%s: failed to reallocate to %u bytes",
--				       "get_next_token()", i);
-+			{
-+				for(i = 0, j = 0; i < len; j++)
-+					if(s[i] != '\\')
-+					{
-+						t[j] = s[i];
-+						i++;
-+					}
-+					else
-+					{
-+						int seqlen;
-+						t[j] = unescape_char(s + i, &seqlen);
-+						i += seqlen;
-+						if (i > len)
-+							break;
-+					}
-+
-+				tmp = realloc(*token, j + 1);
-+				if (tmp != NULL)
-+					t = tmp;
-+				else
-+					syslog(LOG_ERR, "%s: failed to reallocate to %u bytes",
-+					"get_next_token()", j + 1);
-+			}
-+			t[j] = '\0';
-+			*token = t;
- 		}
- 	}
- 
--	/* return the beginning of the next token */
--	if(deli && *end == deli)
--		end++;
--	while(isspace(*end))
--		end++;
--	return (char *) end;
-+	s += len;
-+	if(deli && *s == deli)
-+		s++;
-+	return (char *) s;
- }
- 
- /* read_permission_line()
-  * parse the a permission line which format is :
-- * (deny|allow) [0-9]+(-[0-9]+) ip/mask [0-9]+(-[0-9]+) regex
-+ * (deny|allow) [0-9]+(-[0-9]+)? ip(/mask)? [0-9]+(-[0-9]+)? (regex)?
-  * ip/mask is either 192.168.1.1/24 or 192.168.1.1/255.255.255.0
-  */
- int
- read_permission_line(struct upnpperm * perm,
--                     char * p)
-+                     const char * p)
- {
--	char * q;
--	int n_bits;
--	int i;
-+	unsigned int dot_cnt;
- 
- 	/* zero memory : see https://github.com/miniupnp/miniupnp/issues/652 */
- 	memset(perm, 0, sizeof(struct upnpperm));
- 
--	/* first token: (allow|deny) */
- 	while(isspace(*p))
- 		p++;
-+
-+	/* first token: (allow|deny) */
- 	if(0 == memcmp(p, "allow", 5))
- 	{
- 		perm->type = UPNPPERM_ALLOW;
-@@ -212,133 +291,61 @@ read_permission_line(struct upnpperm * p
- 	{
- 		return -1;
- 	}
--	while(isspace(*p))
--		p++;
-+
-+	p = get_sep(p);
-+	if(!p)
-+		return -1;
- 
- 	/* second token: eport or eport_min-eport_max */
--	if(!isdigit(*p))
-+	p = get_range(p, &perm->eport_min, &perm->eport_max);
-+	if(!p)
- 		return -1;
--	for(q = p; isdigit(*q); q++);
--	if(*q=='-')
--	{
--		*q = '\0';
--		i = atoi(p);
--		if(i > 65535)
--			return -1;
--		perm->eport_min = (u_short)i;
--		q++;
--		p = q;
--		while(isdigit(*q))
--			q++;
--		*q = '\0';
--		i = atoi(p);
--		if(i > 65535)
--			return -1;
--		perm->eport_max = (u_short)i;
--		if(perm->eport_min > perm->eport_max)
--			return -1;
--	}
--	else if(isspace(*q))
--	{
--		*q = '\0';
--		i = atoi(p);
--		if(i > 65535)
--			return -1;
--		perm->eport_min = perm->eport_max = (u_short)i;
--	}
--	else
--	{
-+
-+	p = get_sep(p);
-+	if(!p)
- 		return -1;
--	}
--	p = q + 1;
--	while(isspace(*p))
--		p++;
- 
--	/* third token:  ip/mask */
--	if(!isdigit(*p))
-+	/* third token: ip/mask */
-+	p = get_addr(p, &perm->address, &dot_cnt);
-+	if(!p)
- 		return -1;
--	for(q = p; isdigit(*q) || (*q == '.'); q++);
--	if(*q=='/')
-+
-+	if(*p!='/')
-+		perm->mask.s_addr = 0xffffffffu;
-+	else
- 	{
--		*q = '\0';
--		if(!inet_aton(p, &perm->address))
--			return -1;
--		q++;
--		p = q;
--		while(isdigit(*q))
--			q++;
--		if(*q == '.')
--		{
--			while(*q == '.' || isdigit(*q))
--				q++;
--			if(!isspace(*q))
--				return -1;
--			*q = '\0';
--			if(!inet_aton(p, &perm->mask))
--				return -1;
--		}
--		else if(!isspace(*q))
-+		p++;
-+		p = get_addr(p, &perm->mask, &dot_cnt);
-+		if(!p)
- 			return -1;
--		else
-+		/* inet_aton(): When only one part is given, the value is stored
-+		 * directly in the network address without any byte
-+		 * rearrangement. */
-+		if(!dot_cnt)
- 		{
--			*q = '\0';
--			n_bits = atoi(p);
-+			unsigned int n_bits = ntohl(perm->mask.s_addr);
- 			if(n_bits > 32)
- 				return -1;
--			perm->mask.s_addr = htonl(n_bits ? (0xffffffffu << (32 - n_bits)) : 0);
-+			perm->mask.s_addr = !n_bits ? 0 : htonl(0xffffffffu << (32 - n_bits));
- 		}
- 	}
--	else if(isspace(*q))
--	{
--		*q = '\0';
--		if(!inet_aton(p, &perm->address))
--			return -1;
--		perm->mask.s_addr = 0xffffffffu;
--	}
--	else
--	{
-+
-+	p = get_sep(p);
-+	if(!p)
- 		return -1;
--	}
--	p = q + 1;
- 
- 	/* fourth token: iport or iport_min-iport_max */
--	while(isspace(*p))
--		p++;
--	if(!isdigit(*p))
-+	p = get_range(p, &perm->iport_min, &perm->iport_max);
-+	if(!p)
- 		return -1;
--	for(q = p; isdigit(*q); q++);
--	if(*q=='-')
--	{
--		*q = '\0';
--		i = atoi(p);
--		if(i > 65535)
--			return -1;
--		perm->iport_min = (u_short)i;
--		q++;
--		p = q;
--		while(isdigit(*q))
--			q++;
--		*q = '\0';
--		i = atoi(p);
--		if(i > 65535)
--			return -1;
--		perm->iport_max = (u_short)i;
--		if(perm->iport_min > perm->iport_max)
--			return -1;
--	}
--	else if(isspace(*q) || *q == '\0')
--	{
--		*q = '\0';
--		i = atoi(p);
--		if(i > 65535)
--			return -1;
--		perm->iport_min = perm->iport_max = (u_short)i;
--	}
--	else
--	{
-+
-+	if(iseol(*p) || *p == '#')
-+		goto end;
-+	p = get_sep(p);
-+	if(!p)
- 		return -1;
--	}
--	p = q;
-+	if(iseol(*p) || *p == '#')
-+		goto end;
- 
- 	/* fifth token: (optional) regex */
- 	p = get_next_token(p, &perm->re, 1);
-@@ -381,6 +388,7 @@ read_permission_line(struct upnpperm * p
- 		}
- 	}
- 
-+end:
- #ifdef DEBUG
- 	printf("perm rule added : %s %hu-%hu %08x/%08x %hu-%hu %s\n",
- 	       (perm->type==UPNPPERM_ALLOW) ? "allow" : "deny",
-@@ -427,8 +435,26 @@ write_permlist(int fd, const struct upnp
- 		write(fd, buf, l);
- 		if(perm->re)
- 		{
--			write(fd, " ", 1);
--			write(fd, perm->re, strlen(perm->re));
-+			const char * p;
-+			write(fd, " \"", 2);
-+			for(p = perm->re; *p != '\0'; p++)
-+			{
-+				if(*p == '"')
-+				{
-+					write(fd, "\\\"", 2);
-+					continue;
-+				}
-+
-+				if(*p == '\\')
-+				{
-+					write(fd, p, 1);
-+					p++;
-+					if(*p == '\0')
-+						break;
-+				}
-+				write(fd, p, 1);
-+			}
-+			write(fd, "\"", 1);
- 		}
- 		write(fd, "\n", 1);
- 	}
---- a/upnppermissions.h
-+++ b/upnppermissions.h
-@@ -43,7 +43,7 @@ struct upnpperm {
-  *  deny 0-65535 0.0.0.0/0 0-65535 */
- int
- read_permission_line(struct upnpperm * perm,
--                     char * p);
-+                     const char * p);
- 
- void
- free_permission_line(struct upnpperm * perm);
-@@ -72,4 +72,3 @@ write_permlist(int fd, const struct upnp
- #endif
- 
- #endif
--
diff --git a/net/miniupnpd/patches/030-allow-private-fix.patch b/net/miniupnpd/patches/030-allow-private-fix.patch
deleted file mode 100644
index b8a42b2b56..0000000000
--- a/net/miniupnpd/patches/030-allow-private-fix.patch
+++ /dev/null
@@ -1,18 +0,0 @@
-From 9ff3c71bfb6c7f7bf525d6bda447387de7257aff Mon Sep 17 00:00:00 2001
-From: Vlad Starodubtsev <vlad.starodubtsev@keenetic.com>
-Date: Wed, 6 Aug 2025 19:04:18 +0300
-Subject: [PATCH] miniupnpd: fix ALLOWPRIVATEIPV4MASK condition
-
-Fixed usage of private/reserved WAN addresses if they are allowed in the configuration.
-
---- a/upnpsoap.c
-+++ b/upnpsoap.c
-@@ -370,7 +370,7 @@ GetExternalIPAddress(struct upnphttp * h
- 				ext_if_name);
- 			ext_ip_addr[0] = '\0';
- 		} else if (addr_is_reserved(&addr)) {
--			if (!GETFLAG(ALLOWPRIVATEIPV4MASK)) {
-+			if (GETFLAG(ALLOWPRIVATEIPV4MASK)) {
- 				syslog(LOG_WARNING, "IGNORED : private/reserved address %s is not suitable for external IP", ext_ip_addr);
- 			} else {
- 				syslog(LOG_NOTICE, "private/reserved address %s is not suitable for external IP", ext_ip_addr);
diff --git a/net/miniupnpd/patches/040-improve-logging.patch b/net/miniupnpd/patches/040-improve-logging.patch
index 998df1cb4e..a630ffcbd0 100644
--- a/net/miniupnpd/patches/040-improve-logging.patch
+++ b/net/miniupnpd/patches/040-improve-logging.patch
@@ -1,6 +1,6 @@
---- a/minissdp.c
-+++ b/minissdp.c
-@@ -1274,7 +1274,7 @@ ProcessSSDPData(int s, const char *bufr,
+--- a/minissdp.c	2026-04-01 06:24:16.000000000 +0800
++++ b/minissdp.c	2026-10-16 13:55:54.128940238 +0800
+@@ -1276,7 +1276,7 @@
  					else
  						snprintf(ver_str, sizeof(ver_str), "%d", known_service_types[i].version);
  #endif
@@ -9,7 +9,7 @@
  #ifdef DELAY_MSEARCH_RESPONSE
  					delay = random() / (1 + RAND_MAX / (1000 * mx_value));
  #ifdef DEBUG
-@@ -1303,7 +1303,7 @@ ProcessSSDPData(int s, const char *bufr,
+@@ -1305,7 +1305,7 @@
  #ifdef DELAY_MSEARCH_RESPONSE
  				unsigned int delay_increment = (mx_value * 1000) / 15;
  #endif
@@ -18,7 +18,7 @@
  				for(i=0; known_service_types[i].s; i++)
  				{
  #ifdef DELAY_MSEARCH_RESPONSE
-@@ -1361,7 +1361,7 @@ ProcessSSDPData(int s, const char *bufr,
+@@ -1363,7 +1363,7 @@
  #endif
  				if(0 == memcmp(st, uuidvalue_igd, l))
  				{
@@ -27,7 +27,7 @@
  					SendSSDPResponse(s, sender, st, st_len, "",
  					                 announced_host, http_port,
  #ifdef ENABLE_HTTPS
-@@ -1371,7 +1371,7 @@ ProcessSSDPData(int s, const char *bufr,
+@@ -1373,7 +1373,7 @@
  				}
  				else if(0 == memcmp(st, uuidvalue_wan, l))
  				{
@@ -36,7 +36,7 @@
  					SendSSDPResponse(s, sender, st, st_len, "",
  					                 announced_host, http_port,
  #ifdef ENABLE_HTTPS
-@@ -1381,7 +1381,7 @@ ProcessSSDPData(int s, const char *bufr,
+@@ -1383,7 +1383,7 @@
  				}
  				else if(0 == memcmp(st, uuidvalue_wcd, l))
  				{
@@ -45,14 +45,14 @@
  					SendSSDPResponse(s, sender, st, st_len, "",
  					                 announced_host, http_port,
  #ifdef ENABLE_HTTPS
-@@ -1646,4 +1646,3 @@ SubmitServicesToMiniSSDPD(const char * h
+@@ -1648,4 +1648,3 @@
  	syslog(LOG_DEBUG, "%d service submitted to MiniSSDPd", i);
  	return 0;
  }
 -
---- a/miniupnpd.8
-+++ b/miniupnpd.8
-@@ -42,10 +42,10 @@ and do not filter out low priority messa
+--- a/miniupnpd.8	2025-01-11 06:30:11.000000000 +0800
++++ b/miniupnpd.8	2026-10-16 13:55:54.129091269 +0800
+@@ -42,10 +42,10 @@
  report system uptime instead of daemon uptime to clients.
  .TP
  .B \-S0
@@ -65,7 +65,7 @@
  .TP
  .BI \-u " uuid"
  set the uuid of the UPnP Internet Gateway Device.
-@@ -83,7 +83,7 @@ examples :
+@@ -83,7 +83,7 @@
  sets the value of BOOTID.UPNP.ORG SSDP header
  .TP
  .B \-1
@@ -74,9 +74,9 @@
  .SH "SEE ALSO"
  minissdpd(1) miniupnpc(3)
  .SH BUGS
---- a/miniupnpd.c
-+++ b/miniupnpd.c
-@@ -859,7 +859,7 @@ set_startup_time(void)
+--- a/miniupnpd.c	2026-08-12 07:43:03.000000000 +0800
++++ b/miniupnpd.c	2026-10-16 13:56:32.032770704 +0800
+@@ -867,7 +867,7 @@
  			}
  			else
  			{
@@ -85,30 +85,7 @@
  			}
  			fclose(f);
  			startup_time -= uptime;
-@@ -2098,6 +2098,22 @@ init(int argc, char * * argv, struct run
- 		pidfilename = NULL;
- #endif
- 
-+syslog(LOG_NOTICE, "MiniUPnP daemon version " MINIUPNPD_VERSION " starting, enabled protocols: %s%s%s, ext_ifname=%s BOOTID=%u",
-+	GETFLAG(ENABLEUPNPMASK) ? "UPnP IGD" : "",
-+#ifdef ENABLE_NATPMP
-+	GETFLAG(ENABLEUPNPMASK) && GETFLAG(ENABLENATPMPMASK) ? " & " : "",
-+#ifdef ENABLE_PCP
-+	GETFLAG(ENABLENATPMPMASK) ? "PCP/NAT-PMP" : "",
-+#else
-+	GETFLAG(ENABLENATPMPMASK) ? "NAT-PMP" : "",
-+#endif
-+#else
-+	"", "",
-+#endif
-+	ext_if_name, upnp_bootid);
-+syslog(LOG_INFO, "More information at https://miniupnp.tuxfamily.org/ or http://miniupnp.free.fr/");
-+syslog(LOG_NOTICE, "Extra logging by setting log level to info (-v) or debug (-vv)");
-+
- #ifdef USE_SYSTEMD
- 	if (systemd_flag) {
- 		int r = sd_notify(0,
-@@ -2190,14 +2206,14 @@ print_usage:
+@@ -1309,14 +1309,14 @@
  #if defined(USE_PF) || defined(USE_IPF)
  			"\t-L sets packet log in pf and ipf on.\n"
  #endif
@@ -126,7 +103,7 @@
  #endif
  #endif
  			"\t-B sets bitrates reported by daemon in bits per second.\n"
-@@ -2217,7 +2233,7 @@ print_usage:
+@@ -1336,7 +1336,7 @@
  			"\t  \"deny 0-65535 0.0.0.0/0 0-65535\"\n"
  			"\t-b sets the value of BOOTID.UPNP.ORG SSDP header\n"
  #ifdef IGD_V2
@@ -135,8 +112,31 @@
  #endif
  			"\t-v enables LOG_INFO messages, -vv LOG_DEBUG as well (default with -d)\n"
  			"\t-h / --help prints this help and quits.\n"
-@@ -2398,21 +2414,9 @@ main(int argc, char * * argv)
- 		return 0;
+@@ -2228,6 +2228,22 @@
+ 		pidfilename = NULL;
+ #endif
+ 
++syslog(LOG_NOTICE, "MiniUPnP daemon version " MINIUPNPD_VERSION " starting, enabled protocols: %s%s%s, ext_ifname=%s BOOTID=%u",
++	GETFLAG(ENABLEUPNPMASK) ? "UPnP IGD" : "",
++#ifdef ENABLE_NATPMP
++	GETFLAG(ENABLEUPNPMASK) && GETFLAG(ENABLENATPMPMASK) ? " & " : "",
++#ifdef ENABLE_PCP
++	GETFLAG(ENABLENATPMPMASK) ? "PCP/NAT-PMP" : "",
++#else
++	GETFLAG(ENABLENATPMPMASK) ? "NAT-PMP" : "",
++#endif
++#else
++	"", "",
++#endif
++	ext_if_name, upnp_bootid);
++syslog(LOG_INFO, "More information at https://miniupnp.tuxfamily.org/ or http://miniupnp.free.fr/");
++syslog(LOG_NOTICE, "Extra logging by setting log level to info (-v) or debug (-vv)");
++
+ #ifdef USE_SYSTEMD
+ 	if (systemd_flag) {
+ 		int r = sd_notify(0,
+@@ -2431,21 +2447,9 @@
+ 		goto shutdown;
  	}
  
 -	syslog(LOG_INFO, "version " MINIUPNPD_VERSION " starting%s%sext if %s BOOTID=%u",
@@ -158,25 +158,25 @@
  	}
  #endif
  
-@@ -2456,7 +2460,7 @@ main(int argc, char * * argv)
- 			return 1;
- 		}
- 		v.port = listen_port;
+@@ -2494,7 +2498,7 @@
+ #ifdef ENABLE_NFQUEUE
+ 		nfqueue_data.http_port = listen_port;
+ #endif /* ENABLE_NFQUEUE */
 -		syslog(LOG_NOTICE, "HTTP listening on port %d", v.port);
 +		syslog(LOG_NOTICE, "Listening for UPnP IGD (SOAP/HTTP) traffic on port %d/TCP", v.port);
  #if defined(V6SOCKETS_ARE_V6ONLY) && defined(ENABLE_IPV6)
  		if(!GETFLAG(IPV6DISABLEDMASK))
  		{
-@@ -2482,7 +2486,7 @@ main(int argc, char * * argv)
- 			return 1;
- 		}
- 		v.https_port = listen_port;
+@@ -2525,7 +2529,7 @@
+ #ifdef ENABLE_NFQUEUE
+ 		nfqueue_data.https_port = listen_port;
+ #endif /* ENABLE_NFQUEUE */
 -		syslog(LOG_NOTICE, "HTTPS listening on port %d", v.https_port);
 +		syslog(LOG_NOTICE, "Listening for UPnP IGD (SOAP/HTTPS) traffic on port %d/TCP", v.https_port);
  #if defined(V6SOCKETS_ARE_V6ONLY) && defined(ENABLE_IPV6)
  		shttpsl_v4 =  OpenAndConfHTTPSocket(&listen_port, 0);
  		if(shttpsl_v4 < 0)
-@@ -2496,11 +2500,11 @@ main(int argc, char * * argv)
+@@ -2540,11 +2544,11 @@
  		if(!GETFLAG(IPV6DISABLEDMASK)) {
  			if(find_ipv6_addr(lan_addrs.lh_first ? lan_addrs.lh_first->ifname : NULL,
  			                  ipv6_addr_for_http_with_brackets, sizeof(ipv6_addr_for_http_with_brackets)) > 0) {
@@ -190,7 +190,7 @@
  				SETFLAG(IPV6DISABLEDMASK);
  			}
  		}
-@@ -2571,22 +2575,22 @@ main(int argc, char * * argv)
+@@ -2620,22 +2624,22 @@
  	}
  
  #ifdef ENABLE_NATPMP
@@ -218,7 +218,7 @@
  			       NATPMP_PORT);
  		}
  #endif
-@@ -2695,6 +2699,24 @@ main(int argc, char * * argv)
+@@ -2746,6 +2750,24 @@
  	}
  #endif /* HAS_LIBCAP_NG */
  
@@ -243,7 +243,7 @@
  #ifdef USE_SYSTEMD
  	if (v.systemd_notify) {
  		upnp_update_status();
-@@ -3162,7 +3184,7 @@ main(int argc, char * * argv)
+@@ -3213,7 +3235,7 @@
  		/* process SSDP packets */
  		if(sudp >= 0 && FD_ISSET(sudp, &readset))
  		{
@@ -252,7 +252,7 @@
  #ifdef ENABLE_HTTPS
  			ProcessSSDPRequest(sudp, (unsigned short)v.port, (unsigned short)v.https_port);
  #else
-@@ -3172,7 +3194,7 @@ main(int argc, char * * argv)
+@@ -3223,7 +3245,7 @@
  #ifdef ENABLE_IPV6
  		if(sudpv6 >= 0 && FD_ISSET(sudpv6, &readset))
  		{
@@ -261,9 +261,9 @@
  #ifdef ENABLE_HTTPS
  			ProcessSSDPRequest(sudpv6, (unsigned short)v.port, (unsigned short)v.https_port);
  #else
---- a/miniupnpd.conf
-+++ b/miniupnpd.conf
-@@ -88,7 +88,8 @@
+--- a/miniupnpd.conf	2025-05-26 06:10:20.000000000 +0800
++++ b/miniupnpd.conf	2026-10-16 13:55:54.129544020 +0800
+@@ -89,7 +89,8 @@
  # 120s and 86400s (24h) are suggested values from PCP-base
  #min_lifetime=120
  #max_lifetime=86400
@@ -273,7 +273,7 @@
  #pcp_allow_thirdparty=yes
  
  # table names for netfilter nft. Default is "filter" for both
-@@ -131,8 +132,8 @@
+@@ -132,8 +133,8 @@
  #bitrate_up=1000000
  #bitrate_down=10000000
  
@@ -284,9 +284,9 @@
  #secure_mode=no
  
  # Default presentation URL is HTTP address on port 80
---- a/natpmp.c
-+++ b/natpmp.c
-@@ -231,7 +231,7 @@ void ProcessIncomingNATPMPPacket(int s,
+--- a/natpmp.c	2025-05-05 05:46:14.000000000 +0800
++++ b/natpmp.c	2026-10-16 13:55:54.129630822 +0800
+@@ -231,7 +231,7 @@
  		syslog(LOG_ERR, "inet_ntop(natpmp): %m");
  	}
  
@@ -295,7 +295,7 @@
  	       senderaddrstr, ntohs(senderaddr->sin_port), n);
  
  	if(n<2 || ((((req[1]-1)&~1)==0) && n<12)) {
-@@ -260,7 +260,7 @@ void ProcessIncomingNATPMPPacket(int s,
+@@ -260,7 +260,7 @@
  		resp[3] = 1;	/* unsupported version */
  	} else switch(req[1]) {
  	case 0:	/* Public address request */
@@ -304,7 +304,7 @@
  		FillPublicAddressResponse(resp, senderaddr->sin_addr.s_addr);
  		resplen = 12;
  		break;
-@@ -340,7 +340,7 @@ void ProcessIncomingNATPMPPacket(int s,
+@@ -340,7 +340,7 @@
  						eport_first = eport;
  					} else if(eport == eport_first) { /* no eport available */
  						if(any_eport_allowed == 0) { /* all eports rejected by permissions */
@@ -313,9 +313,9 @@
  							       eport, proto_itoa(proto), senderaddrstr, iport);
  							resp[3] = 2;	/* Not Authorized/Refused */
  						} else { /* at least one eport allowed (but none available) */
---- a/netfilter_nft/nftnlrdr_misc.c
-+++ b/netfilter_nft/nftnlrdr_misc.c
-@@ -104,7 +104,7 @@ nft_mnl_connect(void)
+--- a/netfilter_nft/nftnlrdr_misc.c	2026-08-12 07:43:03.000000000 +0800
++++ b/netfilter_nft/nftnlrdr_misc.c	2026-10-16 13:55:54.129714417 +0800
+@@ -110,7 +110,7 @@
  		return -1;
  	}
  	mnl_portid = mnl_socket_get_portid(mnl_sock);
@@ -324,9 +324,9 @@
  	return 0;
  }
  
---- a/netfilter_nft/nftpinhole.c
-+++ b/netfilter_nft/nftpinhole.c
-@@ -159,7 +159,6 @@ find_pinhole(const char * ifname,
+--- a/netfilter_nft/nftpinhole.c	2025-05-05 05:46:17.000000000 +0800
++++ b/netfilter_nft/nftpinhole.c	2026-10-16 13:55:54.129809022 +0800
+@@ -159,7 +159,6 @@
  		   (0 == memcmp(&daddr, &p->daddr6, sizeof(struct in6_addr)))) {
  
  			if (sscanf(p->desc, PINEHOLE_LABEL_FORMAT_SKIPDESC, &uid, &ts) != 2) {
@@ -334,7 +334,7 @@
  				continue;
  			}
  
-@@ -397,7 +396,6 @@ get_pinhole_info(unsigned short uid,
+@@ -396,7 +395,6 @@
  			if (timestamp) {
  				int uid_temp;
  				if (sscanf(p->desc, PINEHOLE_LABEL_FORMAT_SKIPDESC, &uid_temp, &ts) != 2) {
@@ -342,7 +342,7 @@
  					continue;
  				}
  
-@@ -461,7 +459,6 @@ clean_pinhole_list(unsigned int * next_t
+@@ -460,7 +458,6 @@
  			continue;
  
  		if (sscanf(p->desc, PINEHOLE_LABEL_FORMAT_SKIPDESC, &uid, &ts) != 2) {
@@ -350,9 +350,9 @@
  			continue;
  		}
  
---- a/pcpserver.c
-+++ b/pcpserver.c
-@@ -1087,7 +1087,7 @@ static void CreatePCPMap(pcp_info_t *pcp
+--- a/pcpserver.c	2026-04-07 07:43:40.000000000 +0800
++++ b/pcpserver.c	2026-10-16 13:55:54.129882920 +0800
+@@ -1094,7 +1094,7 @@
  	else
  		r = CreatePCPMap_NAT(pcp_msg_info);
  	pcp_msg_info->result_code = r;
@@ -361,9 +361,9 @@
  	      "PCP MAP: %s mapping %s %hu->%s:%hu '%s'",
  	       r == PCP_SUCCESS ? "added" : "failed to add",
  	       proto_itoa(pcp_msg_info->protocol),
---- a/pf/pfpinhole.c
-+++ b/pf/pfpinhole.c
-@@ -246,12 +246,10 @@ int find_pinhole(const char * ifname,
+--- a/pf/pfpinhole.c	2025-05-05 05:46:17.000000000 +0800
++++ b/pf/pfpinhole.c	2026-10-16 13:55:54.130018973 +0800
+@@ -246,12 +246,10 @@
  		   (0 == memcmp(&daddr, &RULE.dst.addr.v.a.addr.v6, sizeof(struct in6_addr)))) {
  #ifdef USE_LIBPFCTL
  			if(sscanf(RULE.label[0], PINEHOLE_LABEL_FORMAT_SKIPDESC, &uid, &ts) != 2) {
@@ -376,7 +376,7 @@
  				continue;
  			}
  #endif /* USE_LIBPFCTL */
-@@ -563,7 +561,6 @@ int clean_pinhole_list(unsigned int * ne
+@@ -565,7 +563,6 @@
  			return -1;
  		}
  		if(sscanf(RULE.label[0], PINEHOLE_LABEL_FORMAT_SKIPDESC, &uid, &ts) != 2) {
@@ -384,7 +384,7 @@
  			continue;
  		}
  #else /* USE_LIBPFCTL */
-@@ -574,7 +571,6 @@ int clean_pinhole_list(unsigned int * ne
+@@ -576,7 +573,6 @@
  			return -1;
  		}
  		if(sscanf(RULE.label, PINEHOLE_LABEL_FORMAT_SKIPDESC, &uid, &ts) != 2) {
@@ -392,9 +392,9 @@
  			continue;
  		}
  #endif /* USE_LIBPFCTL */
---- a/upnphttp.c
-+++ b/upnphttp.c
-@@ -513,7 +513,7 @@ ProcessHTTPPOST_upnphttp(struct upnphttp
+--- a/upnphttp.c	2026-10-16 13:55:54.119869149 +0800
++++ b/upnphttp.c	2026-10-16 13:55:54.130136591 +0800
+@@ -517,7 +517,7 @@
  		if(h->req_soapActionOff > 0)
  		{
  			/* we can process the request */
@@ -403,7 +403,7 @@
  			       h->req_soapActionLen, h->req_buf + h->req_soapActionOff);
  			ExecuteSoapAction(h,
  				h->req_buf + h->req_soapActionOff,
-@@ -818,7 +818,7 @@ ProcessHttpQuery_upnphttp(struct upnphtt
+@@ -822,7 +822,7 @@
  	for(i = 0; i<15 && *p != '\r'; i++)
  		HttpVer[i] = *(p++);
  	HttpVer[i] = '\0';
@@ -412,9 +412,9 @@
  	       h->clientaddr_str, HttpCommand, HttpUrl, HttpVer);
  	ParseHttpHeaders(h);
  	if(h->req_HostOff > 0 && h->req_HostLen > 0) {
---- a/upnpredirect.c
-+++ b/upnpredirect.c
-@@ -641,7 +641,7 @@ get_upnp_rules_state_list(int max_rules_
+--- a/upnpredirect.c	2026-08-12 07:43:03.000000000 +0800
++++ b/upnpredirect.c	2026-10-16 13:55:54.130248258 +0800
+@@ -648,7 +648,7 @@
  	{
  		if(tmp->to_remove)
  		{
@@ -423,9 +423,9 @@
  			       tmp->eport, proto_itoa(tmp->proto));
  			_upnp_delete_redir(tmp->eport, tmp->proto);
  			*p = tmp->next;
---- a/upnpsoap.c
-+++ b/upnpsoap.c
-@@ -847,7 +847,7 @@ GetSpecificPortMappingEntry(struct upnph
+--- a/upnpsoap.c	2026-10-16 13:55:54.120209962 +0800
++++ b/upnpsoap.c	2026-10-16 13:55:54.130355798 +0800
+@@ -913,7 +913,7 @@
  	}
  	else
  	{
@@ -434,7 +434,7 @@
  		       action,
  		       r_host ? r_host : "NULL", ext_port, protocol, int_ip,
  		       (unsigned int)iport, desc, leaseduration);
-@@ -1112,7 +1112,7 @@ GetGenericPortMappingEntry(struct upnpht
+@@ -1196,7 +1196,7 @@
  		return;
  	}
  
@@ -443,7 +443,7 @@
  
  	rhost[0] = '\0';
  	r = upnp_get_redirection_infos_by_index((int)index, &eport, protocol, &iport,
-@@ -2486,4 +2486,3 @@ SoapError(struct upnphttp * h, int errCo
+@@ -2709,4 +2709,3 @@
  	BuildResp2_upnphttp(h, 500, "Internal Server Error", body, bodylen);
  	SendRespAndClose_upnphttp(h);
  }
